contracts/dmarket.js

```diff
@@ -0,0 +1,1258 @@
+  buyer,
+  seller,
+  symbol,
+  quoteToken,
+  qtPrecision,
+  quantity,
+  price,
+  volume,
+  buyTxId,
+  sellTxId) => {
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestampSec = blockDate.getTime() / 1000;
+  const timestampMinus24hrs = blockDate.setUTCDate(blockDate.getUTCDate() - 1) / 1000;
+
+  // clean history
+  let tradesToDelete = await api.db.find(
+    'tradesHistory',
+    {
+      symbol,
+      quoteToken,
+      timestamp: {
+        $lt: timestampMinus24hrs,
+      },
+    },
+  );
+  let nbTradesToDelete = tradesToDelete.length;
+
+  while (nbTradesToDelete > 0) {
+    for (let index = 0; index < nbTradesToDelete; index += 1) {
+      const trade = tradesToDelete[index];
+      await updateVolumeMetric(trade.symbol, trade.quoteToken, qtPrecision, trade.volume, false);
+      await api.db.remove('tradesHistory', trade);
+    }
+    tradesToDelete = await api.db.find(
+      'tradesHistory',
+      {
+        symbol,
+        quoteToken,
+        timestamp: {
+          $lt: timestampMinus24hrs,
+        },
+      },
+    );
+    nbTradesToDelete = tradesToDelete.length;
+  }
+
+  // add order to the history
+  const newTrade = {};
+  newTrade.type = type;
+  newTrade.buyer = buyer;
+  newTrade.seller = seller;
+  newTrade.symbol = symbol;
+  newTrade.quoteToken = quoteToken;
+  newTrade.quantity = quantity;
+  newTrade.price = price;
+  newTrade.timestamp = timestampSec;
+  newTrade.volume = volume;
+  newTrade.buyTxId = buyTxId;
+  newTrade.sellTxId = sellTxId;
+  await api.db.insert('tradesHistory', newTrade);
+  await updatePriceMetrics(symbol, quoteToken, qtPrecision, price);
+};
+
+const removeExpiredOrders = async (table) => {
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestampSec = blockDate.getTime() / 1000;
+
+  // clean orders
+  let nbOrdersToDelete = 0;
+  let ordersToDelete = await api.db.find(
+    table,
+    {
+      expiration: {
+        $lte: timestampSec,
+      },
+    },
+  );
+
+  nbOrdersToDelete = ordersToDelete.length;
+  while (nbOrdersToDelete > 0) {
+    for (let index = 0; index < nbOrdersToDelete; index += 1) {
+      const order = ordersToDelete[index];
+      let quantity;
+      let symbol;
+
+      if (table === 'buyBook') {
+        symbol = order.quoteToken;
+        quantity = order.tokensLocked;
+      } else {
+        // eslint-disable-next-line prefer-destructuring
+        symbol = order.symbol;
+        // eslint-disable-next-line prefer-destructuring
+        quantity = order.quantity;
+      }
+
+      // unlock tokens
+      await api.transferTokens(order.account, symbol, quantity, 'user');
+
+      await api.db.remove(table, order);
+
+      if (table === 'buyBook') {
+        api.emit('orderExpired', { type: 'buy', txId: order.txId });
+
+        await updateAskMetric(order.symbol);
+      } else {
+        api.emit('orderExpired', { type: 'sell', txId: order.txId });
+
+        await updateBidMetric(order.symbol);
```

Comment:
updateBidMetric, updateAskMetric appear to need a quoteToken as parameter. (Test coverage to catch this?)

Reply:
updated, also added test units for expired orders